File: src/test/java/com/digitalpetri/strictmachine/dsl/ActionBuilderTest.java
package com.digitalpetri.strictmachine.dsl;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.ArrayList;
import org.junit.jupiter.api.Test;

class ActionBuilderTest {

  @Test
  void actionsAreExecutedInOrder() throws InterruptedException {
    var fb = new FsmBuilder<State, Event>();

    var executed = new ArrayList<Integer>();

    fb.when(State.S1)
        .on(Event.E1.class)
        .transitionTo(State.S2)
        .execute(ctx -> executed.add(0))
        .executeLast(ctx -> executed.add(1))
        .executeFirst(ctx -> executed.add(2));

    fb.build(State.S1).fireEventBlocking(new Event.E1());

    assertEquals(2, (int) executed.get(0));
    assertEquals(0, (int) executed.get(1));
    assertEquals(1, (int) executed.get(2));
  }

}

File: src/test/java/com/digitalpetri/strictmachine/dsl/ActionBuilderTest.kt
package com.digitalpetri.strictmachine.dsl

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class ActionBuilderTest {

    @Test
    fun `Actions are executed in order`() {
        val fb = FsmBuilder<State, Event>()

        val executed = mutableListOf<Int>()

        fb.`when`(State.S1)
            .on(Event.E1)
            .transitionTo(State.S2)
            .execute { executed.add(0) }
            .executeLast { executed.add(1) }
            .executeFirst { executed.add(2) }

        fb.build(State.S1).fireEventBlocking(Event.E1)

        Assertions.assertTrue(executed[0] == 2)
        Assertions.assertTrue(executed[1] == 0)
        Assertions.assertTrue(executed[2] == 1)
    }

}
